src/video/SDL_video.c

```diff
@@ -1488,108 +1420,143 @@ static int SDL_UpdateFullscreenMode(SDL_Window *window, SDL_bool fullscreen)
-            return 0;
+        if (SDL_SetDisplayModeForDisplay(display, mode) < 0) {
+            return -1;
         }
-    }
+        if (_this->SetWindowFullscreen) {
+            _this->SetWindowFullscreen(_this, window, display, SDL_TRUE);
+        }
+        display->fullscreen_window = window;
 
-    /* See if there are any fullscreen windows */
-    for (other = _this->windows; other; other = other->next) {
-        SDL_bool setDisplayMode = SDL_FALSE;
+#if defined(__ANDROID__)
+        /* Android may not resize the window to exactly what our fullscreen mode is,
+         * especially on windowed Android environments like the Chromebook or Samsung DeX.
+         * Given this, we shouldn't use the mode size. As such, Android's SetWindowFullscreen
+         * will generate the window event for us with the proper final size.
+         */
+#elif defined(__WIN32__)
+        /* This is also unnecessary on Win32 (WIN_SetWindowFullscreen calls SetWindowPos,
+         * WM_WINDOWPOSCHANGED will send SDL_EVENT_WINDOW_RESIZED).
+         */
+#else
+        if (mode) {
+            mode_w = mode->screen_w;
+            mode_h = mode->screen_h;
+        } else {
+            mode_w = display->desktop_mode.screen_w;
+            mode_h = display->desktop_mode.screen_h;
+        }
+        if (window->w != mode_w || window->h != mode_h) {
+            resized = SDL_TRUE;
+        }
+#endif
+        if (resized) {
+            SDL_SendWindowEvent(window, SDL_EVENT_WINDOW_RESIZED, mode_w, mode_h);
+        } else {
+            SDL_OnWindowResized(window);
+        }
+
+        /* Restore the cursor position */
+        SDL_RestoreMousePosition(window);
+
+    } else {
 
-        if (other == window) {
-            setDisplayMode = fullscreen;
-        } else if (SDL_WINDOW_FULLSCREEN_VISIBLE(other) &&
-                   SDL_GetVideoDisplayForWindow(other) == display) {
-            setDisplayMode = SDL_TRUE;
+        if (display->fullscreen_window == window) {
+            /* Restore the desktop mode */
+            SDL_SetDisplayModeForDisplay(display, NULL);
+        }
+
+        if (_this->SetWindowFullscreen) {
+            _this->SetWindowFullscreen(_this, window, display, SDL_FALSE);
+        }
+        if (display->fullscreen_window == window) {
+            display->fullscreen_window = NULL;
         }
 
-        if (setDisplayMode) {
-            SDL_DisplayMode fullscreen_mode;
+        SDL_OnWindowResized(window);
 
-            SDL_zero(fullscreen_mode);
+        /* Restore the cursor position */
+        SDL_RestoreMousePosition(window);
+    }
 
-            if (SDL_GetWindowDisplayMode(other, &fullscreen_mode) == 0) {
-                SDL_bool resized = SDL_TRUE;
+    window->last_fullscreen_flags = window->flags;
 
-                if (other->w == fullscreen_mode.screen_w && other->h == fullscreen_mode.screen_h) {
-                    resized = SDL_FALSE;
-                }
+    return retval;
```

Comment:
With the rewritten logic in this function, retval is never set aside from it's declaration, so this can just be
```suggestion
    return 0;
```
and the retval variable can go away.

Reply:
Fixed, thanks!